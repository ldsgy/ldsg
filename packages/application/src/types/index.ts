export * from "./handler-resource-settings";
