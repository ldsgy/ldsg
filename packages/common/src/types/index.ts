export * from "./app-data";
export * from "./handler-service-settings";
export * from "./service-record";
