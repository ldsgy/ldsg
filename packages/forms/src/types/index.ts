export * from "./get-form-info";
export * from "./specific-resource-settings";
