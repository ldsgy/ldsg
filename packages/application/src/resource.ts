import { Resource } from "@ldsg/resource";
import express, { Express } from "express";
import { ApplicationSpecificResourceSettings } from "./types";

type CreateApp = () => Promise<Express>;

interface AddRoutesParams {
  app: Express;
}

type AddRoutes = (params: AddRoutesParams) => Promise<void>;

export class ApplicationResource extends Resource<ApplicationSpecificResourceSettings> {
  addRoutes: AddRoutes = async (params) => {
    const { app } = params;

    const {
      settings: { name },
      getFilteredResources,
    } = this;

    const getFilteredResourcesRes = getFilteredResources({});

    const { resources } = getFilteredResourcesRes;

    resources.forEach((resource) => {
      if ("addRoute" in resource && typeof resource.addRoute === "function") {
        resource?.addRoute();
      }
    });

    app.get("/", (req, res) => {
      res.send(`Hello, ${name ?? "World"}!`);
    });
  };

  createApp: CreateApp = async () => {
    const app = express();

    await this.addRoutes({
      app,
    });

    return app;
  };
}
