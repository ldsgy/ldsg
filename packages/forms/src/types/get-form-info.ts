import { PlatformsParams } from "@ldsg/field-type";
import { ObjectInfo } from "@ldsg/object";
import { GeneralResourceSettings } from "@ldsg/types";
import { WorkflowInfo } from "@ldsg/workflow";

export interface FormInfo extends GeneralResourceSettings {
  /**
   * Form ID
   */
  id: string;

  /**
   * Form Name
   */
  name: string;

  /**
   * Input Object Info
   */
  inputObjectInfo: ObjectInfo;

  /**
   * Output Object Info
   */
  outputObjectInfo: ObjectInfo;

  /**
   * Workflow Info
   */
  workflowInfo: WorkflowInfo;
}

export type GetFormInfo = (params?: PlatformsParams) => FormInfo;
