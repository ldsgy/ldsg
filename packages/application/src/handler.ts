import { Handler } from "@ldsg/handler";
import { Manifest } from "@ldsg/resource";
import { ApplicationResource } from "./resource";
import { ApplicationResourceSettings } from "./types";

export const handler: Handler<
  [Manifest.Resource<ApplicationResourceSettings>]
> = (params) => {
  return new ApplicationResource(params);
};
