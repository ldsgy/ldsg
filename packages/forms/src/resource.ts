import { FormResource } from "@ldsg/form";
import { Resource } from "@ldsg/resource";
import { FormsSpecificResourceSettings } from "./types";

export class FormsResource extends Resource<FormsSpecificResourceSettings> {
  getSubFormResources = () => {
    const { id, getFilteredResources } = this;

    const { resources } = getFilteredResources({
      parentId: id,
    });

    return resources as FormResource[];
  };
}
