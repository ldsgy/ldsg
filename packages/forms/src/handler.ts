import { Handler, HandlerExtendedResourceSettings } from "@ldsg/handler";
import {
  InstantiateResourceParams,
  InstantiateResourceRes,
} from "@ldsg/resource";
import { instantiateResource } from "./instantiate-resource";
import { FormResource } from "./resource";
import { FormSpecificResourceSettings } from "./types";

export const handler: Handler<
  [
    InstantiateResourceParams<
      HandlerExtendedResourceSettings<FormSpecificResourceSettings>
    >
  ],
  InstantiateResourceRes<FormResource>
> = instantiateResource;
