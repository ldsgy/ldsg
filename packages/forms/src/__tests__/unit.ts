import { FieldTypeResource } from "@ldsg/field-type";
import { GraphqlResource } from "@ldsg/graphql";
import { Handler, HandlerResource } from "@ldsg/handler";
import { ObjectResource } from "@ldsg/object";
import { ObjectFieldResource } from "@ldsg/object-field";
import { Resource } from "@ldsg/resource";
import { SpecificResourceSettings } from "@ldsg/types";
import { printSchema } from "graphql";
import { ObjectTypeComposerFieldConfigAsObjectDefinition } from "graphql-compose";
import { FormsResource } from "../resource";
import { FormInfo, GetFormInfo } from "../types";

interface ASpecificResourceSettings extends SpecificResourceSettings {
  /**
   * Form Name
   */
  name: string;

  /**
   * Input Object Resource Id
   */
  inputObjectResourceId: string;

  /**
   * Output Object Resource Id
   */
  outputObjectResourceId: string;

  /**
   * Workflow Resource Id
   */
  workflowResourceId: string;
}

class AResource extends Resource<ASpecificResourceSettings> {
  getFormInfo: GetFormInfo = (params) => {
    const { platform } = params;

    const {
      settings: { title, description, name },
      getResourcesFromSettings,
    } = this;

    const { inputObjectResource, outputObjectResource, workflowResource } =
      getResourcesFromSettings();

    const { objectInfo: inputObjectInfo } = (
      inputObjectResource as ObjectResource
    ).getObjectInfo({ platform });

    const { objectInfo: outputObjectInfo } = (
      outputObjectResource as ObjectResource
    ).getObjectInfo({ platform });

    const res: FormInfo = {
      title,
      description,
      name,
      inputObjectInfo,
      outputObjectInfo,
    };

    return res;
  };
}

const handler: Handler<
  [
    {
      /**
       * Field Properties
       */
      fieldProperties?: any;
      /**
       * Platform
       * Such as mongoose\formily.
       */
      platform: string;
    }
  ],
  {}
> = (params) => {
  const { fieldProperties, platform } = params;

  const { max } = fieldProperties;

  let res;

  switch (platform) {
    case "graphql": {
      const type: ObjectTypeComposerFieldConfigAsObjectDefinition<
        any,
        any
      >["type"] = "String";

      res = {
        type,
      };

      break;
    }

    case "mongoose": {
      res = {
        type: "String",
        ...(max
          ? {
              maxLength: max,
            }
          : {}),
      };

      break;
    }

    default: {
      res = {};

      break;
    }
  }

  return res;
};

test("forms", () => {
  const graphqlResource = new GraphqlResource({
    id: "test-graphql",
    kind: "GRAPHQL",
    parentId: "test-application",
    settings: {
      title: "测试GraphQL",
      description: "",
      code: "",
      dependencies: [],
    },
  });

  new HandlerResource({
    id: "test-handler",
    kind: "HANDLER",
    parentId: "root",
    settings: {
      title: "测试处理程序",
      description: "",
      code: "",
      dependencies: [],
    },
    handler,
  });

  new FieldTypeResource({
    id: "test-field-type",
    kind: "FIELD_TYPE",
    parentId: "root",
    settings: {
      title: "文本",
      description: "可用来存储各种文本",
      fieldPropertiesSchema: {
        type: "object",
        properties: {
          max: {
            type: "integer",
            title: "配置最长字符",
            description: "长度不可以超过此值",
          },
        },
      },
      handlerResourceId: "test-handler",
    },
  });

  new ObjectFieldResource({
    id: "test-object-1-field-1",
    kind: "OBJECT_FIELD",
    parentId: "test-object-1",
    settings: {
      title: "测试对象1测试字段1",
      description: "",
      name: "test-1",
      fieldTypeResourceId: "test-field-type",
      properties: {
        max: "10",
      },
    },
  });

  new ObjectFieldResource({
    id: "test-object-1-field-2",
    kind: "OBJECT_FIELD",
    parentId: "test-object-1",
    settings: {
      title: "测试对象1测试字段2",
      description: "",
      name: "test-2",
      fieldTypeResourceId: "test-field-type",
      properties: {
        max: "10",
      },
    },
  });

  new ObjectFieldResource({
    id: "test-object-1-field-3",
    kind: "OBJECT_FIELD",
    parentId: "test-object-1",
    settings: {
      title: "测试对象1测试字段3",
      description: "",
      name: "test-3",
      fieldTypeResourceId: "test-field-type",
      properties: {
        max: "10",
      },
    },
  });

  new ObjectResource({
    id: "test-object-1",
    kind: "OBJECT",
    parentId: "root",
    settings: {
      title: "测试对象1",
      description: "",
      name: "test-object-1",
    },
  });

  new ObjectFieldResource({
    id: "test-object-2-field-1",
    kind: "OBJECT_FIELD",
    parentId: "test-object-2",
    settings: {
      title: "测试对象2测试字段1",
      description: "",
      name: "test-1",
      fieldTypeResourceId: "test-field-type",
      properties: {
        max: "10",
      },
    },
  });

  new ObjectResource({
    id: "test-object-2",
    kind: "OBJECT",
    parentId: "root",
    settings: {
      title: "测试对象2",
      description: "",
      name: "test-object-2",
    },
  });

  const formsResource = new FormsResource({
    id: "test-forms",
    kind: "FORM",
    parentId: "test-application",
    settings: {
      title: "测试表单组",
      description: "",
    },
  });

  new AResource({
    id: "test-a-1",
    kind: "FORM",
    parentId: "test-forms",
    settings: {
      title: "测试表单1",
      description: "",
      name: "test-a-1",
      inputObjectResourceId: "test-object-1",
      outputObjectResourceId: "test-object-2",
      workflowResourceId: "",
    },
  });

  new AResource({
    id: "test-a-2",
    kind: "FORM",
    parentId: "test-forms",
    settings: {
      title: "测试表单2",
      description: "",
      name: "test-a-2",
      inputObjectResourceId: "test-object-2",
      outputObjectResourceId: "test-object-1",
      workflowResourceId: "",
    },
  });

  const getFormInfoListRes = formsResource.getFormInfoList({
    platform: "mongoose",
  });

  expect(getFormInfoListRes).toMatchSnapshot();

  const { schema } = graphqlResource.getGraphQLSchema();

  const schemaDSL = printSchema(schema);

  console.log(schemaDSL);

  expect(schemaDSL).toMatchSnapshot();
});
