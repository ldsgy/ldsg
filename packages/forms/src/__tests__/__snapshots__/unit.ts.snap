// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`forms 1`] = `
{
  "formInfoList": [
    {
      "description": "",
      "inputObjectInfo": {
        "description": "",
        "fieldInfoList": [
          {
            "description": "",
            "name": "test-1",
            "title": "测试对象1测试字段1",
            "typeInfo": {
              "maxLength": "10",
              "type": "String",
            },
          },
          {
            "description": "",
            "name": "test-2",
            "title": "测试对象1测试字段2",
            "typeInfo": {
              "maxLength": "10",
              "type": "String",
            },
          },
          {
            "description": "",
            "name": "test-3",
            "title": "测试对象1测试字段3",
            "typeInfo": {
              "maxLength": "10",
              "type": "String",
            },
          },
        ],
        "name": "test-object-1",
        "title": "测试对象1",
      },
      "name": "test-a-1",
      "outputObjectInfo": {
        "description": "",
        "fieldInfoList": [
          {
            "description": "",
            "name": "test-1",
            "title": "测试对象2测试字段1",
            "typeInfo": {
              "maxLength": "10",
              "type": "String",
            },
          },
        ],
        "name": "test-object-2",
        "title": "测试对象2",
      },
      "title": "测试表单1",
      "workflowInfo": {
        "execute": [Function],
      },
    },
    {
      "description": "",
      "inputObjectInfo": {
        "description": "",
        "fieldInfoList": [
          {
            "description": "",
            "name": "test-1",
            "title": "测试对象2测试字段1",
            "typeInfo": {
              "maxLength": "10",
              "type": "String",
            },
          },
        ],
        "name": "test-object-2",
        "title": "测试对象2",
      },
      "name": "test-a-2",
      "outputObjectInfo": {
        "description": "",
        "fieldInfoList": [
          {
            "description": "",
            "name": "test-1",
            "title": "测试对象1测试字段1",
            "typeInfo": {
              "maxLength": "10",
              "type": "String",
            },
          },
          {
            "description": "",
            "name": "test-2",
            "title": "测试对象1测试字段2",
            "typeInfo": {
              "maxLength": "10",
              "type": "String",
            },
          },
          {
            "description": "",
            "name": "test-3",
            "title": "测试对象1测试字段3",
            "typeInfo": {
              "maxLength": "10",
              "type": "String",
            },
          },
        ],
        "name": "test-object-1",
        "title": "测试对象1",
      },
      "title": "测试表单2",
      "workflowInfo": {
        "execute": [Function],
      },
    },
  ],
}
`;

exports[`forms 2`] = `
"type Query {
  """测试表单1"""
  testA1(
    """测试对象1测试字段1"""
    test1: String

    """测试对象1测试字段2"""
    test2: String

    """测试对象1测试字段3"""
    test3: String
  ): TestObject2

  """测试表单2"""
  testA2(
    """测试对象2测试字段1"""
    test1: String
  ): TestObject1
}

"""测试对象2"""
type TestObject2 {
  """测试对象2测试字段1"""
  test1: String
}

"""测试对象1"""
type TestObject1 {
  """测试对象1测试字段1"""
  test1: String

  """测试对象1测试字段2"""
  test2: String

  """测试对象1测试字段3"""
  test3: String
}

type Mutation {
  """测试表单1"""
  testA1(
    """测试对象1测试字段1"""
    test1: String

    """测试对象1测试字段2"""
    test2: String

    """测试对象1测试字段3"""
    test3: String
  ): TestObject2

  """测试表单2"""
  testA2(
    """测试对象2测试字段1"""
    test1: String
  ): TestObject1
}"
`;
